feat: Allow passing arbitrary Tensorizer serialization and deserialization kwargs; update docs (#7)
coreweave/vllm: docs/models/extensions/tensorizer.md

vLLM supports loading models with [CoreWeave's Tensorizer](https://docs.coreweave.com/coreweave-machine-learning-and-ai/inference/tensorizer).
vLLM model tensors that have been serialized to disk, an HTTP/HTTPS endpoint, or S3 endpoint can be deserialized
at runtime extremely quickly directly to the GPU, resulting in significantly
lower Pod startup times and CPU memory usage. Tensor encryption is also supported.

> vLLM fully integrates Tensorizer in to its model loading machinery. The

Reviewer: ```suggestion
vLLM fully integrates Tensorizer in its model loading machinery. The
```

at runtime extremely quickly directly to the GPU, resulting in significantly

> shorter Pod startup times and CPU memory usage. Tensor encryption is also supported.

Reviewer: This wasn't added as part of this PR, but this should be rephrased while you're at it.
```suggestion
lower Pod startup times and CPU memory usage. Tensor encryption is also supported.
```

vLLM fully integrates Tensorizer in to its model loading machinery. The
following will give a brief overview on how to get started with using
Tensorizer on vLLM.

## The basics
To load a model using Tensorizer, it first needs to be serialized by Tensorizer.
The example script in [examples/others/tensorize_vllm_model.py](https://github.com/vllm-project/vllm/blob/main/examples/others/tensorize_vllm_model.py)
takes care of this process.

The core frontend object of note integrating Tensorizer is 
`TensorizerConfig`, defined [here](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/model_loader/tensorizer.py#L135-L214)
It's a config object holding important state and passed to any serialization 

> or deserialization operation. When loading with Tensorizer using the vLLM 

Reviewer: Suggesting to change this mainly for succinctness but also so that it doesn't link to a specific line number range on the main branch, since those line numbers will quickly become inaccurate. Also, it doesn't look like that class is actually stateful (i.e. it looks basically immutable), so I'm suggesting dropping the reference to "state" here. If that's not accurate, you can add that part back.
```suggestion
The `TensorizerConfig` class is used to customize Tensorizer's behaviour,
defined in [vllm/model_executor/model_loader/tensorizer.py](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/model_loader/tensorizer.py).
It is passed to any serialization or deserialization operation.
When loading with Tensorizer using the vLLM 
```

library rather than through a model-serving entrypoint, it gets passed to 
the `LLM` entrypoint class directly. Here's an example of loading a model

> saved at `"s3://my-bucket/vllm/facebook/opt-125m/v1/model.tensors"`.

Reviewer: ```suggestion
saved at `"s3://my-bucket/vllm/facebook/opt-125m/v1/model.tensors"`:
```

```python
from vllm import LLM
from vllm.model_executor.model_loader.tensorizer import TensorizerConfig

path_to_tensors = "s3://my-bucket/vllm/facebook/opt-125m/v1/model.tensors"

model_ref = "facebook/opt-125m"
tensorizer_config = TensorizerConfig(
        tensorizer_uri=path_to_tensors,

>     )

Reviewer: Indentation issue
```suggestion
    tensorizer_uri=path_to_tensors
)
```

llm = LLM(
    model_ref,
    load_format="tensorizer",
    model_loader_extra_config=tensorizer_config,
)
```

But that code won't work unless you actually have your serialized model 
tensors `model.tensors`, so let's walk through a basic example of serializing 

> `facebook/opt-125m` using the example script, and then loading it for inference.

Reviewer: Nitpick copyediting for professionalism
```suggestion
However, the above code will not function until you have successfully
serialized the model tensors with Tensorizer to get the `model.tensors`
file shown. The following section walks through an end-to-end example
of serializing `facebook/opt-125m` with the example script,
and then loading it for inference.
```

## Saving a vLLM model with Tensorizer
To save a model with Tensorizer, call the example script with the necessary
CLI arguments. The docstring for the script itself explains the CLI args
and how to use it properly in great detail, and we'll use one of the
examples from the docstring directly, assuming we want to save our model at
our S3 bucket example `s3://my-bucket`:

```bash
python examples/others/tensorize_vllm_model.py \
   --model facebook/opt-125m \
   serialize \
   --serialized-directory s3://my-bucket \
   --suffix v1
```

This saves the model tensors at 
`s3://my-bucket/vllm/facebook/opt-125m/v1/model.tensors`.

## Serving the model using Tensorizer
Once the model is serialized where you want it, you need to pass
`--load-format=tensorizer` as well as a JSON string to the
`--model-loader-extra-config` CLI arg for `vllm serve`, specifying
all the keyword arguments you'd normally pass to the `TensorizerConfig`
initializer, as if you were instantiating it like this, for instance:
`TensorizerConfig(**{"tensorizer_uri": "foo"})`. Here's an example bash
script that neatly lays out the JSON string before passing it to

> `--model-loader-extra-config`:

Reviewer: Clarity
```suggestion
Once the model is serialized where you want it, you must specify the
following additional command line parameters to `vllm serve` to load the model:
- `--load-format=tensorizer`
- `--model-loader-extra-config`
  - This expects a JSON string specifying keyword arguments for
    the `TensorizerConfig` object
  - For example, `--model-loader-extra-config='{"tensorizer_uri": "foo"}'`
    would be equivalent to `TensorizerConfig(**{"tensorizer_uri": "foo"})`
The shell snippet below shows a typical invocation of `vllm serve`
with these parameters specified:
```

```bash
#!/bin/bash

read -r -d '' JSON << EOF
{
  "tensorizer_uri": "s3://my-bucket/vllm/facebook/opt-125m/v1/model.tensors",
  "stream_kwargs": {"force_http": "False"},
  "deserialization_kwargs": {"verify_hash": "True", "num_readers": 8}
}
EOF

> MODEL_LOADER_EXTRA_CONFIG=$(echo "$JSON" | tr -d '\n')

Reviewer: Reduce unnecessary complexity here. Also, booleans were being specified as strings like `"False"` instead of proper JSON booleans. If `"False"` is deserialized with `json.load`, it will end up as a truthy value in Python, which isn't what you want here.

Please also update this in the PR description.
```suggestion
MODEL_LOADER_EXTRA_CONFIG='{
  "tensorizer_uri": "s3://my-bucket/vllm/facebook/opt-125m/v1/model.tensors",
  "stream_kwargs": {"force_http": false},
  "deserialization_kwargs": {"verify_hash": true, "num_readers": 8}
}'
```

vllm serve facebook/opt-125m \
  --load-format=tensorizer \
  --model-loader-extra-config="$MODEL_LOADER_EXTRA_CONFIG"
```

Note in this case, if the directory to the model artifacts at
`s3://my-bucket/vllm/facebook/opt-125m/v1/` doesn't at least have a `config.
json` file, you'll want to pass `facebook/opt-125m` as the model tag like

> it was done in the example script above. In our example, we just added a

Reviewer: Can you rephrase this slightly for clarity? Maybe something like "the `facebook/opt-125m` positional argument in the above example is required when there is no `config.json` file at [...]"

Author: Yes -- I'll update it to account for the changes in the sister PR as well.

```

This saves the model tensors at 
`s3://my-bucket/vllm/facebook/opt-125m/v1/model.tensors`, as well as all other 
artifacts needed to load the model at that same directory.

## Serving the model using Tensorizer
Once the model is serialized where you want it, all that is needed is to 
pass that directory with the model artifacts to `vllm serve` in the case above, 
one can simply do:

```bash

> #!/bin/bash

Reviewer: A shebang is excessive for a one line shell command.
```suggestion
```

vllm serve s3://my-bucket/vllm/facebook/opt-125m/v1 --load-format=tensorizer
```

Please note that authentication to S3 is still required. Tensorizer will 
automatically look for a `~/.s3cfg` or `~/.aws/config` and `~/.aws/credentials`
file locally to authenticate, but also can be given the environment variables 
`AWS_ACCESS_KEY_ID` or `S3_ACCESS_KEY_ID` for the object storage access key,
`AWS_SECRET_ACCESS_KEY` or `S3_SECRET_ACCESS_KEY` for the object storage 

> secret key, and `AWS_S3_ENDPOINT_URL` or `S3_ENDPOINT_URL` for the endpoint url.

Reviewer: 1. Prefer to call it "object storage."
2. `AWS_S3_ENDPOINT_URL` is not the name of the environment variable (it's `AWS_ENDPOINT_URL_S3` or `AWS_ENDPOINT_URL`), but either way, we should just list the `AWS_*` ones since those are the only ones supported in Tensorizer itself, not any of the `S3_*` ones.
3. We should also note somewhere here that any normal [environment variables for configuring Boto3](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#using-environment-variables) will work.
```suggestion
Please note that object storage authentication is still required.
To authenticate, Tensorizer searches for a `~/.s3cfg` configuration file
([s3cmd](https://s3tools.org/kb/item14.htm)'s format),
or `~/.aws/config` and `~/.aws/credentials` files (`boto3` and the
`aws` CLI's format), but authentication can also be configured using
[any normal `boto3` environment variables](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#using-environment-variables),
such as `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`,
and `AWS_ENDPOINT_URL_S3`.
```

If only the model tensors are saved, you can still provide that as the only 
artifact in your directory to load from, and vLLM will fetch the rest using the 
`--model-loader-extra-config` CLI arg, passing a JSON string with the args 

> for `TensorizerConfig`.

Reviewer: What does "vLLM will fetch the rest using the `--model-loader-extra-config` CLI arg" mean? Doesn't it fetch the rest using the positional argument for the model tag, not `--model-loader-extra-config`?

Author: Yes, it expects the model tag to be a HuggingFace model ID. I'll make sure this is clear. I've done a poor job explaining that here.

```bash

> #!/bin/bash

Reviewer: Note that this is POSIX `sh`-compatible
```suggestion
#!/bin/sh
```
